src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala

```diff
@@ -131,37 +187,19 @@ class IpLookups(
    * @return Tuple containing the results of the
    *         LookupServices
    */
-  private def performLookupsWithoutLruCache(ip: String): IpLookupResult = {
-
-    val ipAddress = getIpAddress(ip)
-
-    /**
-     * Creates a Either boxing the result of using a lookup service on the ip
-     * @param service ISP, domain or connection type LookupService
-     * @return the result of the lookup
-     */
-    def getLookup(service: Option[SpecializedReader]): Option[Either[Throwable, String]] =
-      service.map { s =>
-        for {
-          ipA <- ipAddress
-          v   <- s.getValue(ipA)
-        } yield v
-      }
-
-    val ipLocation: Option[Either[Throwable, IpLocation]] =
-      geoService.map { gs =>
-        for {
-          ipA <- ipAddress
-          v   <- Either.catchNonFatal(gs.city(ipA))
-        } yield IpLocation.apply(v)
-      }
-
-    IpLookupResult(
-      ipLocation,
-      getLookup(ispService),
-      getLookup(orgService),
-      getLookup(domainService),
-      getLookup(connectionTypeService)
+  private def performLookupsWithoutLruCache[F[_]: Sync](ip: String): F[IpLookupResult] = {
```

Comment:
@chuwy I managed to switch the library to `Sync` but it doesn't look like `for { ... } yield` syntax works anymore. I get `value flatMap is not a member of type parameter F[com.snowplowanalytics.maxmind.iplookups.model.IpLookupResult]` when I try it.

The `flatMap` syntax is a bit ugly so I'm wondering if there's a way to get `for yield` back?

Reply:
Also, I think you mentioned on the call there was a way to get synchronization using `Sync`. The code isn't thread safe right now since `LruMap` isn't synchronized. I'm wondering what the best way to make it thread safe is?

Reply:
regarding the for-comp it's just a matter of

```scala
import cats.syntax.flatMap._
import cats.syntax.functor._
```

regarding synchronization, wrapping operations in an effect should take care of that.

Reply:
I had a weird problem recently where even with `cats.syntax` it couldn't resolve proper `map` (and therefore `flatMap`) methods for `for`-comprehension. Adding explicit implicit (pun) of a `Functor` into scope helped:

```
  def dosomething(process: Process)
                (implicit S: Sync[F]): F[Unit] = {
    implicit val F: Functor[F] = S    // Only this helped
    for {
      list <- someAction
      _ <- list.traverse[F, Data](process)
    } yield ()
  }
```

Maybe it was a matter of `traverse`, but sometimes unfortunately it is a matter of playing with implicit scope.

> regarding synchronization, wrapping operations in an effect should take care of that

Yep, agreed.

Reply:
ah, it might just be:

```scala
def do[F[_]: Sync : Monad](): F[Unit] = for {
} yield ()
```